cpp/src/parquet/schema.cc

```diff
@@ -243,6 +291,55 @@ void PrimitiveNode::VisitConst(Node::ConstVisitor* visitor) const {
 // ----------------------------------------------------------------------
 // Group node
 
+GroupNode::GroupNode(const std::string& name, Repetition::type repetition,
+                     const NodeVector& fields, LogicalType::type logical_type, int id)
+    : Node(Node::GROUP, name, repetition, logical_type, id), fields_(fields) {
+  // For forward compatibility, create an equivalent logical annotation
+  logical_annotation_ = LogicalAnnotation::FromConvertedType(logical_type_);
+  DCHECK(logical_annotation_ &&
+         (logical_annotation_->is_nested() || logical_annotation_->is_none()) &&
+         logical_annotation_->is_compatible(logical_type_));
+
+  field_name_to_idx_.clear();
+  auto field_idx = 0;
+  for (NodePtr& field : fields_) {
+    field->SetParent(this);
+    field_name_to_idx_.emplace(field->name(), field_idx++);
+  }
+}
+
+GroupNode::GroupNode(const std::string& name, Repetition::type repetition,
+                     const NodeVector& fields,
+                     std::shared_ptr<const LogicalAnnotation> logical_annotation, int id)
+    : Node(Node::GROUP, name, repetition, logical_annotation, id), fields_(fields) {
+  if (logical_annotation_) {
+    // Check for annotation type <=> node type consistency
+    if (logical_annotation_->is_nested()) {
+      // For backward compatibility, assign equivalent legacy converted type (if possible)
+      logical_type_ = logical_annotation_->ToConvertedType(NULLPTR);
```

Comment:
you can use `nullptr` here, NULLPTR is only necessary in header files

Reply:
Changed in b6ccf8c9612ed37877153c456a755e3f69231c02.